components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java

```
@@ -2381,6 +2425,28 @@ private String doUserAuthorization(OAuthMessage oAuthMessage, String sessionData
         }
     }
 
+    /**
+     * Validate scopes before consent page
+     *
+     * @param  oAuthMessage oAuthMessage
+     * @param oauth2Params oauth2Params
+     */
+    private void handleAuthorizationBeforeConsent(OAuthMessage oAuthMessage, OAuth2Parameters oauth2Params,
+                                             OAuth2AuthorizeReqDTO authzReqDTO)
+            throws IdentityOAuth2ScopeValidationException {
+
+        try {
+            AuthorizationHandlerManager authzHandlerManager = AuthorizationHandlerManager.getInstance();
+            OAuthAuthzReqMessageContext authzReqMsgCtx = authzHandlerManager.
+                    handleAuthorizationBeforeConsent(authzReqDTO);
+            addOAuthAuthzReqMessageContextTOSessionDataCacheEntry(oAuthMessage, authzReqMsgCtx);
+            oauth2Params.setScopes(new HashSet<>(Arrays.asList(authzReqMsgCtx.getApprovedScope())));
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            log.error("Error occurred when validating scopes");
```

Throw an exception here instead of just logging an error